src/elastio-snap.c

```diff
@@ -2813,14 +2869,47 @@ static int cow_init(struct snap_device *dev, const char *path, uint64_t elements
 	cm->nr_changed_blocks = 0;
 	cm->flags = 0;
 	cm->allocated_sects = 0;
-	cm->file_max = file_max;
 	cm->sect_size = sect_size;
 	cm->seqid = seqid;
-	cm->log_sect_pages = get_order(sect_size*8);
+
+	// Perhaps the code below needs a small explanation
+	//
+	// The COW file structure is as follows:
+	// +--------------------+----------+------+
+	// | HEADER (+padding)  | SECTIONS | DATA |
+	// +--------------------+----------+------+
+	//
+	// +--------------------------------------+
+	// |               SECTION                |
+	// +--------------------------------------+
+	// | block 1 | block 2 | ... | block 4096 |
+	// +--------------------------------------+
+	//
+	// +--------------------------------------+
+	// |                BLOCK                 |
+	// +--------------------------------------+
+	// |              0...4096 bytes          |
+	// +--------------------------------------+
+	//
+	// 1. Each sector has 4096 cow block mappings
+	// 2. Each mapping describes 4096 bytes segment
+	//
+	// Suppose we have 256 MB disk.
+	// Taking PAGE_SIZE == COW_BLOCK_SIZE == 4096 bytes, yield:
+	//
+	//  - 65536 blocks to be mapped;
+	//  - We need 16 sectors to map everything
+	//
+	//  Belowe we calculate that.
+	//
+	//  Here '8' is sizeof(uint64_t) (please refer to 'struct cow_section')
+
+	cm->log_sect_pages = get_order(sect_size * 8);
 	cm->total_sects = NUM_SEGMENTS(elements, cm->log_sect_pages + PAGE_SHIFT - 3);
 	cm->allowed_sects = __cow_calculate_allowed_sects(cache_size, cm->total_sects);
 	cm->data_offset = COW_HEADER_SIZE + (cm->total_sects * (sect_size*8));
 	cm->curr_pos = cm->data_offset / COW_BLOCK_SIZE;
+	cm->file_max = file_max + cm->data_offset;
```

Comment:
If `fallocate` is not specified, we calculate the cow file size as 10% of the disk volume. If we do that, we actually want it to be, well, 10% instead of 10% - memory for section storage. The section storage isn't huge, can be like 1-50 MB depending on disk capacity. But we promised 10% for **data** to be stored and I wouldn't like to implicitly steal some of this memory.